### PR TITLE
fix(autoloading): Add missing autoloaders for shipped apps

### PR DESCRIPTION
**Pull request by ChristophWurst**
* Resolves: n/a

## Summary

Shipped apps in this repo have their own authorative (optimized) composer autoloader. But not all of them. The apps still work, but we need a dynamic PSR4 autoloader that has to access the filesystem to check for class existence. Furthermore I'm assuming that with each app autoloader we add, there is a chance that multiple autoloaders try to load a certain file.

By adding the few missing autoloaders I was able to reduce `php occ` runtime by 1%. `GET <host>/status.php` improved from 330.20 requests per second to 344.20 request per second or 4.2%.

The autoloader was missing for the following apps
* ~~dashboard~~ https://github.com/nextcloud/server/pull/36440
* ~~files_external~~ https://github.com/nextcloud/server/pull/36469
* theming https://github.com/nextcloud/server/pull/36652
* ~~weather_status~~ https://github.com/nextcloud/server/pull/36444

I assume that distributed Nextcloud setups with the php code on shared storage will see an even bigger performance improvement.

## 100x`php occ` benchmark

```time for i in `seq 1 100`; do php occ; done```

<details>

### 100x occ before

real    0m50,214s
user    0m43,513s
sys     0m6,638s

### 100x occ after

real    0m49,866s
user    0m43,296s
sys     0m6,529s
</details>

## Apache Bench

``ab -n 10000 -c 20 https://localhost/status.php``

<details>

### Before
```
$ ab -n 10000 -c 20 https://localhost/status.php
This is ApacheBench, Version 2.3 <$Revision: 1901567 $>
Copyright 1996 Adam Twiss, Zeus Technology Ltd, http://www.zeustech.net/
Licensed to The Apache Software Foundation, http://www.apache.org/

Benchmarking localhost (be patient)
Completed 1000 requests
Completed 2000 requests
Completed 3000 requests
Completed 4000 requests
Completed 5000 requests
Completed 6000 requests
Completed 7000 requests
Completed 8000 requests
Completed 9000 requests
Completed 10000 requests
Finished 10000 requests


Server Software:        nginx/1.22.1
Server Hostname:        localhost
Server Port:            443
SSL/TLS Protocol:       TLSv1.2,ECDHE-RSA-AES256-GCM-SHA384,4096,256
Server Temp Key:        X25519 253 bits
TLS Server Name:        localhost

Document Path:          /status.php
Document Length:        174 bytes

Concurrency Level:      20
Time taken for tests:   30.285 seconds
Complete requests:      10000
Failed requests:        0
Total transferred:      14089962 bytes
HTML transferred:       1740000 bytes
Requests per second:    330.20 [#/sec] (mean)
Time per request:       60.570 [ms] (mean)
Time per request:       3.029 [ms] (mean, across all concurrent requests)
Transfer rate:          454.34 [Kbytes/sec] received

Connection Times (ms)
              min  mean[+/-sd] median   max
Connect:        7   16   8.4     13     134
Processing:    19   44  15.2     41     170
Waiting:       19   44  15.2     41     169
Total:         28   60  16.2     57     282

Percentage of the requests served within a certain time (ms)
  50%     57
  66%     62
  75%     66
  80%     68
  90%     77
  95%     88
  98%    108
  99%    121
 100%    282 (longest request)
```

### After
```
$ ab -n 10000 -c 20 https://localhost/status.php
This is ApacheBench, Version 2.3 <$Revision: 1901567 $>
Copyright 1996 Adam Twiss, Zeus Technology Ltd, http://www.zeustech.net/
Licensed to The Apache Software Foundation, http://www.apache.org/

Benchmarking localhost (be patient)
Completed 1000 requests
Completed 2000 requests
Completed 3000 requests
Completed 4000 requests
Completed 5000 requests
Completed 6000 requests
Completed 7000 requests
Completed 8000 requests
Completed 9000 requests
Completed 10000 requests
Finished 10000 requests


Server Software:        nginx/1.22.1
Server Hostname:        localhost
Server Port:            443
SSL/TLS Protocol:       TLSv1.2,ECDHE-RSA-AES256-GCM-SHA384,4096,256
Server Temp Key:        X25519 253 bits
TLS Server Name:        localhost

Document Path:          /status.php
Document Length:        174 bytes

Concurrency Level:      20
Time taken for tests:   29.053 seconds
Complete requests:      10000
Failed requests:        0
Total transferred:      14090218 bytes
HTML transferred:       1740000 bytes
Requests per second:    344.20 [#/sec] (mean)
Time per request:       58.106 [ms] (mean)
Time per request:       2.905 [ms] (mean, across all concurrent requests)
Transfer rate:          473.62 [Kbytes/sec] received

Connection Times (ms)
              min  mean[+/-sd] median   max
Connect:        7   15   7.3     12      88
Processing:    17   43  13.8     40     143
Waiting:       17   43  13.9     40     143
Total:         28   58  14.1     55     213

Percentage of the requests served within a certain time (ms)
  50%     55
  66%     59
  75%     63
  80%     65
  90%     73
  95%     84
  98%    102
  99%    112
 100%    213 (longest request)
```
</details>

## Checklist

- Code is [properly formatted](https://docs.nextcloud.com/server/latest/developer_manual/digging_deeper/continuous_integration.html#linting)
- [Sign-off message](https://github.com/src-d/guide/blob/master/developer-community/fix-DCO.md) is added to all commits
- [x] Tests ([unit](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#unit-tests), [integration](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#integration-tests), api and/or acceptance) are included
- [x] Screenshots before/after for front-end changes
- [x] Documentation ([manuals](https://github.com/nextcloud/documentation/) or wiki) has been updated or is not required
- [x] [Backports requested](https://github.com/nextcloud/backportbot/#usage) where applicable (ex: critical bugfixes)

### CODE REVIEW

**Review by PVince81 — APPROVED** [review]
:+1: nice catch

any reason why our build autoloader script did not cover those ? otherwise CI would have caught this discrepancy, would be good to add a step there for that

**Comment by ChristophWurst** [comment]
> any reason why our build autoloader script did not cover those ? otherwise CI would have caught this discrepancy, would be good to add a step there for that

``if [[ -e ${app}/composer/composer.json ]]; then``

Only apps that have the composer file are checked

**Review by juliusknorr — APPROVED** [review]

**Review by juliusknorr — CHANGES_REQUESTED** [review]
files_external seems to have some issues still according to CI:

Migration step 'OCA\Files_External\Migration\Version1011Date20200630192246' is unknown
[15](https://github.com/nextcloud/server/actions/runs/3900761480/jobs/6661803873#step:5:16)

**Review by st3iny — APPROVED** [review]

**Comment by ChristophWurst** [comment]
> otherwise CI would have caught this discrepancy, would be good to add a step there for that

Added

**Comment by ChristophWurst** [comment]
> files_external seems to have some issues still according to CI:
> 
> Migration step 'OCA\Files_External\Migration\Version1011Date20200630192246' is unknown [15](https://github.com/nextcloud/server/actions/runs/3900761480/jobs/6661803873#step:5:16)

Typo in the PSR4 spec

File: .github/workflows/s3-external.yml [hunk 1]
```diff
@@ -53,7 +53,7 @@ jobs:
         run: |
           mkdir data
           ./occ maintenance:install --verbose --database=sqlite --database-name=nextcloud --database-host=127.0.0.1 --database-user=root --database-pass=rootpassword --admin-user admin --admin-pass password
-          ./occ app:enable --force ${{ env.APP_NAME }}
+          ./occ app:enable -vvv --force ${{ env.APP_NAME }}
```
Comment (tcitworld) [inline .github/workflows/s3-external.yml:56]: **nitpick:** This doesn't seem related.

File: core/Command/App/Enable.php [hunk 2]
```diff
@@ -121,9 +121,6 @@ private function enableApp(string $appId, array $groupIds, bool $forceEnable, Ou
 		} catch (AppPathNotFoundException $e) {
 			$output->writeln($appId . ' not found');
 			$this->exitCode = 1;
-		} catch (\Exception $e) {
-			$output->writeln($e->getMessage());
-			$this->exitCode = 1;
```
Comment (tcitworld) [inline core/Command/App/Enable.php:126]: **nitpick:** This doesn't seem related.

Comment (ChristophWurst) [reply core/Command/App/Enable.php:126]: This will be dropped as well

File: .github/workflows/s3-external.yml (hunk 1 above)
Comment (ChristophWurst) [reply .github/workflows/s3-external.yml:56]: I will drop https://github.com/nextcloud/server/pull/36112/commits/a7928205951a7db0a9624167c4b5f8417093be04 with the final rebase

File: build/autoloaderchecker.sh [hunk 3]
```diff
@@ -22,13 +22,17 @@ echo "Regenerating main autoloader"
 $COMPOSER_COMMAND dump-autoload -d $REPODIR
 
 for app in ${REPODIR}/apps/*; do
+	git check-ignore ${app} -q
+	if git check-ignore ${app} -q ; then
+		echo
+		echo "${app} is not shipped. Ignoring autoloader regeneration:w"
```
Comment (nickvergessen) [inline build/autoloaderchecker.sh:28]:
```suggestion
		echo "${app} is not shipped. Ignoring autoloader regeneration"
```

Trying to close an editor or something?

**Review by nickvergessen — APPROVED** [review]

**Comment by ChristophWurst** [comment]
CI is not happy about the theming app changes :s

**Pull request closed without merging**